### PR TITLE
[Merged by Bors] - feat: port Data.List.Sort

### PR DESCRIPTION
**Pull request by Ruben-VandeVelde**
---

[Open in Gitpod](https://gitpod.io/from-referrer/)

### CODE REVIEW

File: Mathlib/Data/List/Sort.lean [hunk 1]
```diff
@@ -0,0 +1,464 @@
...
+  monotone_iff_of_fn_sorted.1 h
+#align list.monotone.of_fn_sorted List.Monotone.of_fn_sorted
+
+end Monotone
+
+section sort
+
+variable {α : Type uu} (r : α → α → Prop) [DecidableRel r]
+
+-- mathport name: «expr ≼ »
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Sort.lean:151]:
```suggestion
```

File: Mathlib/Data/List/Sort.lean [hunk 2]
```diff
@@ -0,0 +1,464 @@
...
+
+/-- A tuple is monotone if and only if the list obtained from it is sorted. -/
+theorem monotone_iff_of_fn_sorted : Monotone f ↔ (ofFn f).Sorted (· ≤ ·) := by
+  simp_rw [Sorted, pairwise_iff_get, length_ofFn, get_ofFn, monotone_iff_forall_lt]
+  refine ⟨fun h i j hij => h <| Fin.mk_lt_mk.mpr hij, fun h ⟨i, hi⟩ ⟨j, hj⟩ hij => ?_⟩
+  exact h ⟨i, (length_ofFn f).symm ▸ hi⟩ ⟨j, (length_ofFn f).symm ▸ hj⟩ hij
+#align list.monotone_iff_of_fn_sorted List.monotone_iff_of_fn_sorted
+
+/-- The list obtained from a monotone tuple is sorted. -/
+theorem Monotone.of_fn_sorted (h : Monotone f) : (ofFn f).Sorted (· ≤ ·) :=
```
Comment (jcommelin) [inline Mathlib/Data/List/Sort.lean:141]:
```suggestion
theorem monotone_iff_ofFn_sorted : Monotone f ↔ (ofFn f).Sorted (· ≤ ·) := by
  simp_rw [Sorted, pairwise_iff_get, length_ofFn, get_ofFn, monotone_iff_forall_lt]
  refine ⟨fun h i j hij => h <| Fin.mk_lt_mk.mpr hij, fun h ⟨i, hi⟩ ⟨j, hj⟩ hij => ?_⟩
  exact h ⟨i, (length_ofFn f).symm ▸ hi⟩ ⟨j, (length_ofFn f).symm ▸ hj⟩ hij
#align list.monotone_iff_of_fn_sorted List.monotone_iff_of_fn_sorted

/-- The list obtained from a monotone tuple is sorted. -/
theorem Monotone.ofFn_sorted (h : Monotone f) : (ofFn f).Sorted (· ≤ ·) :=
```

File: Mathlib/Data/List/Sort.lean [hunk 3]
```diff
@@ -0,0 +1,464 @@
...
+  | [a] => [a]
+  | a :: b :: l => by
+    -- Porting note: rewrote to make `mergeSort_cons_cons` proof easier
+    let l₁ := (split (a :: b :: l)).1
+    let l₂ := (split (a :: b :: l)).2
+    have e : split (a :: b :: l) = ⟨l₁, l₂⟩ := rfl
+    have h := length_split_lt e
+    have := h.1
+    have := h.2
+    exact merge r (mergeSort l₁) (mergeSort l₂)
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Sort.lean:346]: I'm not sure this is the right solution - perhaps someone who knows about algorithms in lean4 could take a look. Maybe @digama0?

File: Mathlib/Data/List/Sort.lean [hunk 4]
```diff
@@ -0,0 +1,464 @@
...
+section TotalAndTransitive
+
+variable {r} [IsTotal α r] [IsTrans α r]
+
+theorem Sorted.merge : ∀ {l l' : List α}, Sorted r l → Sorted r l' → Sorted r (merge r l l')
+  | [], [], _, _ => by simp [List.merge]
+  | [], b :: l', _, h₂ => by simpa [List.merge] using h₂
+  | a :: l, [], h₁, _ => by simpa [List.merge] using h₁
+  | a :: l, b :: l', h₁, h₂ => by
+    by_cases a ≼ b
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Sort.lean:397]:
```suggestion
    by_cases h : a ≼ b
```

(I wish this didn't work)

File: Mathlib/Data/List/Sort.lean [hunk 5]
```diff
@@ -0,0 +1,464 @@
...
+theorem monotone_iff_of_fn_sorted : Monotone f ↔ (ofFn f).Sorted (· ≤ ·) := by
+  simp_rw [Sorted, pairwise_iff_get, length_ofFn, get_ofFn, monotone_iff_forall_lt]
+  refine ⟨fun h i j hij => h <| Fin.mk_lt_mk.mpr hij, fun h ⟨i, hi⟩ ⟨j, hj⟩ hij => ?_⟩
+  exact h ⟨i, (length_ofFn f).symm ▸ hi⟩ ⟨j, (length_ofFn f).symm ▸ hj⟩ hij
+#align list.monotone_iff_of_fn_sorted List.monotone_iff_of_fn_sorted
+
+/-- The list obtained from a monotone tuple is sorted. -/
+theorem Monotone.of_fn_sorted (h : Monotone f) : (ofFn f).Sorted (· ≤ ·) :=
+  monotone_iff_of_fn_sorted.1 h
+#align list.monotone.of_fn_sorted List.Monotone.of_fn_sorted
```
Comment (Ruben-VandeVelde) [inline Mathlib/Data/List/Sort.lean:143]:
```suggestion
  monotone_iff_ofFn_sorted.1 h
#align list.monotone.of_fn_sorted List.Monotone.ofFn_sorted
```

**Pull request closed without merging**